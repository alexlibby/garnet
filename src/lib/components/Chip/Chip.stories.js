import Chip from "./Chip.svelte";
import { BADGE } from "storybook-addon-badges";

const BADGES = {
  ...BADGE,
  ALPHA: "Alpha",
};

export default {
  title: "Garnet UI Library/Navigation Components/Chip",
  component: Chip,
  props: {
    close: false,
    outline: false,
    chipContent: "Test chip",
  },
};

export let Default = (args) => ({
  Component: Chip,
  props: {
    ...args,
    close: false,
    chipContent: "Test chip",
  },
});

Default.parameters = {
  badges: [BADGES.ALPHA],
};

export let CloseButton = (args) => ({
  Component: Chip,
  props: {
    ...args,
    close: true,
    chipContent: "Test chip",
  },
});

CloseButton.parameters = {
  badges: [BADGES.ALPHA],
};

export let Outline = (args) => ({
  Component: Chip,
  props: {
    ...args,
    outline: true,
    chipContent: "Test chip",
  },
});

Outline.parameters = {
  badges: [BADGES.ALPHA],
};
