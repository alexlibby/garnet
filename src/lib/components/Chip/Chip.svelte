<svelte:options customElement="garnet-chip" />

<script>
  import DeleteIcon from './DeleteIcon.svelte';
  import { createEventDispatcher } from 'svelte';

  export let active = true;
  export let close = false;
  export let chipContent = "";
  export let selected = false;
  export let outline = false;

  const dispatch = createEventDispatcher();

  function onClose(e) {
    active = false;
    dispatch('close', e);
  }

  $: outline = outline;
</script>

<div class="garnet-chip">
  {#if active}
    <!-- svelte-ignore a11y-no-static-element-interactions -->
    <!-- svelte-ignore a11y-click-events-have-key-events -->
    <div
      class="chip"
      class:selected
      class:outline
      data-testid="chip"
      >
      <span>{chipContent}</span>
      {#if close}
        <span class="close" on:click={onClose}>
          <DeleteIcon />
        </span>
      {/if}
      </div>
  {/if}
</div>

<style>
  .garnet-chip {
    display: flex;
    font-family: Arial, Helvetica, sans-serif;
  }
  
  .chip {
    color: #733635;
    align-items: center;
    cursor: default;
    display: inline-flex;
    line-height: 20px;
    max-width: 100%;
    outline: none;
    overflow: hidden;
    padding: 0 12px;
    position: relative;
    text-decoration: none;
    vertical-align: middle;
    white-space: nowrap;
    border-radius: 25px;
    padding: 10px;
    background-color: #d3d3d3;
  }

  .close {
    cursor: pointer;
    margin-left: 6px;
    margin-right: -6px;
    display: flex;
  }

  .close:focus, .close:hover, .close:active {
    opacity: 0.72;
  }

  .outline {
    background-color: #ffffff;
    border: 1px solid #d3d3d3;
  }
</style>
