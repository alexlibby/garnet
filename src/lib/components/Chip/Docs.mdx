import { Canvas, Meta } from "@storybook/addon-docs";
import { ArgTypes } from "@storybook/blocks";
import * as ChipStories from "./Chip.stories";

<Meta of={ChipStories} />

# Chip

Button is the primary component. It has three possible states.

- [Default](#default)
- [CloseButton](#closebutton)
- [Outline](#outline)

## Default

This is the default state for the Chip component:

<Canvas of={ChipStories.Default} />

## CloseButton

This shows the button with a close option:

<Canvas of={ChipStories.CloseButton} />

## Outline

This variant shows the Chip with an outline style:

<Canvas of={ChipStories.Outline} />

## Properties of component

Below is a list of arguments available for this component:

<ArgTypes of={ChipStories} />
