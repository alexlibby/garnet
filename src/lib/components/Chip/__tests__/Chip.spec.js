import { describe, test, expect } from "vitest";
import { axe } from "vitest-axe";
import * as AxeMatchers from "vitest-axe/matchers";
import { render, screen, fireEvent } from "@testing-library/svelte";
import Chip from "../../Chip/Chip.svelte";

const ChipHTML = `<div class="garnet s-clkxn-jNAPHQ"><div class="chip s-clkxn-jNAPHQ" data-testid="chip"><span class="s-clkxn-jNAPHQ">Test chip</span> </div></div>`;
expect.extend(AxeMatchers);

describe("Tests for Chip", () => {
  const props = {
    active: true,
    close: false,
    chipContent: "",
    selected: false,
    outline: false,
  };

  test("should render properly", () => {
    const result = render(Chip, { props });
    expect(() => result).not.toThrow();
  });

  test("show a chip with close icon", () => {
    const result = render(Chip, { ...props, close: true });
    expect(result).not.toBeNull();
  });

  test("should close the chip when clicked", async () => {
    render(Chip, { props });
    const ChipTag = screen.getByTestId("chip");
    await fireEvent.click(ChipTag);
    expect(ChipTag).toBeUndefined;
  });

  test("get a snapshot of component", () => {
    // Create a new container for the test and mount component
    const host = document.createElement("div");
    document.body.appendChild(host);
    const instance = new Chip({ target: host, props: props });

    // check that all rendered OK
    expect(instance).toBeTruthy();

    // Take snapshot
    expect(host.innerHTML).toMatchSnapshot();
  });

  test("should demonstrate no issues with accessibility", async () => {
    const render = () => ChipHTML;
    const result = render();

    const results = await axe(result, {
      rules: {
        region: { enabled: false },
      },
    });

    expect(results).toHaveNoViolations();
  });
});
