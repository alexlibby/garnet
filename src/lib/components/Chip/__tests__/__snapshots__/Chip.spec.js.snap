// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for Chip > get a snapshot of component 1`] = `"<div class="garnet-chip svelte-99jo0f"><div class="chip svelte-99jo0f" data-testid="chip"><span></span> </div></div>"`;
